### PR TITLE
Add PS signatures

### PR DESCRIPTION
**Pull request by lovesh**
Signed-off-by: lovesh <lovesh.bond@gmail.com>

### CODE REVIEW

**Comment by dhh1128** [comment]
I am eager to see this merged, because it will be helpful for some enhancements to Hyperledger Indy credential schemes.

File: libzmix/ps/src/pok_vc.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    // XXX: Error for VC should be independent of PS
+    use crate::errors::PSError;
+    use amcl_wrapper::field_elem::{FieldElement, FieldElementVector};
+    use amcl_wrapper::group_elem::{GroupElement, GroupElementVector};
+    use amcl_wrapper::group_elem_g1::{G1Vector, G1};
+    use amcl_wrapper::group_elem_g2::{G2Vector, G2};
+
+    #[test]
+    fn test_PoK_VC_G1() {
+        // Proof of knowledge of committed values in a vector commitment. The committment lies in group G1.
```
Comment (danintel) [inline libzmix/ps/src/pok_vc.rs]: s/committment/commitment/

File: libzmix/ps/src/pok_vc.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            ProverCommittedG1,
+            ProofG1,
+            G1,
+            G1Vector
+        );
+    }
+
+    #[test]
+    fn test_PoK_VC_G2() {
+        // Proof of knowledge of committed values in a vector commitment. The committment lies in group G2.
```
Comment (danintel) [inline libzmix/ps/src/pok_vc.rs]: s/committment/commitment/

File: libzmix/ps/src/signature.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    /// 1 or more messages are captured in a commitment `commitment`. The remaining known messages are in `messages`.
+    /// This is a blind signature.
+    pub fn new_with_committed_messages(
+        commitment: &SignatureGroup,
+        messages: &[FieldElement],
+        sigkey: &Sigkey,
+        verkey: &Verkey,
+    ) -> Result<Self, PSError> {
+        verkey.validate()?;
+        // There should be commitment to atleast one message
```
Comment (danintel) [inline libzmix/ps/src/signature.rs]: s/atleast/at least/

File: libzmix/ps/src/pok_vc.rs (hunk 1 above)
Comment (lovesh) [reply libzmix/ps/src/pok_vc.rs]: Fixed.

File: libzmix/ps/src/pok_vc.rs (hunk 2 above)
Comment (lovesh) [reply libzmix/ps/src/pok_vc.rs]: Fixed.

File: libzmix/ps/src/signature.rs (hunk 3 above)
Comment (lovesh) [reply libzmix/ps/src/signature.rs]: Fixed.

File: libzmix/ps/src/signature.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        verkey.validate()?;
+        // There should be commitment to at least one message
+        if messages.len() >= verkey.Y.len() {
+            return Err(PSError::UnsupportedNoOfMessages {
+                expected: messages.len(),
+                given: verkey.Y.len(),
+            });
+        }
+
+        let u = FieldElement::random();
```
[the remark this answers is not in the document]
Comment (lovesh) [reply libzmix/ps/src/signature.rs]: The idea is that the signer calls `new_with_committed_messages` after he has verified the proof that commitment is valid. [This code block](https://github.com/hyperledger/ursa/pull/46/files#diff-3b79c2d1f6accb7035421822bd62d6c6R27) in the test shows that signer verifies the proof of knowledge before calling `new_with_committed_messages`

File: libzmix/ps/tests/scenario.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+use ps::keys::keygen;
+use ps::pok_sig::*;
+use ps::signature::Signature;
+use ps::{OtherGroupVec, SignatureGroup};
+use std::collections::{HashMap, HashSet};
+
+#[test]
+fn test_scenario_1() {
+    // User request signer to sign 10 messages where signer knows only 8 messages, the rest 2 are given in a form of commitment.
+    // Once user gets the signature, it engages in a proof of knowledge of signature with a verifier.
```
[the remark this answers is not in the document]
Comment (lovesh) [reply libzmix/ps/tests/scenario.rs]: Regarding `creates a signature proof of knowledge with a verifier`, I am not seeing a difference in the meaning conveyed. `Proof of knowledge of signature` is easier to understand though.

File: libzmix/ps/tests/scenario.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+use amcl_wrapper::group_elem::{GroupElement, GroupElementVector};
+use ps::keys::keygen;
+use ps::pok_sig::*;
+use ps::signature::Signature;
+use ps::{OtherGroupVec, SignatureGroup};
+use std::collections::{HashMap, HashSet};
+
+#[test]
+fn test_scenario_1() {
+    // User request signer to sign 10 messages where signer knows only 8 messages, the rest 2 are given in a form of commitment.
```
[the remark this answers is not in the document]
Comment (lovesh) [reply libzmix/ps/tests/scenario.rs]: Fixed

File: libzmix/ps/src/signature.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply libzmix/ps/src/signature.rs]: I don't follow. What practice?

**Comment by lovesh** [comment]
> There needs to be more sad path tests implemented

I added more.

**Comment by mac-arrap** [comment]
Thank you @lovesh . I will merge

**Pull request merged**